compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParserErrorHandler.java

```diff
@@ -2592,18 +2457,46 @@ private Result seekMatchInExpressionRhs(int lookahead, int currentDepth, int cur
             throw new IllegalStateException("seekMatchInExpressionRhs found: " + parentCtx);
         }
 
-        alternatives = new ParserRuleContext[] { ParserRuleContext.BINARY_OPERATOR, ParserRuleContext.IS_KEYWORD,
-                ParserRuleContext.DOT, ParserRuleContext.ANNOT_CHAINING_TOKEN,
-                ParserRuleContext.OPTIONAL_CHAINING_TOKEN, ParserRuleContext.CONDITIONAL_EXPRESSION,
-                ParserRuleContext.XML_NAVIGATE_EXPR, ParserRuleContext.MEMBER_ACCESS_KEY_EXPR,
-                ParserRuleContext.RIGHT_ARROW, ParserRuleContext.SYNC_SEND_TOKEN, nextContext };
+        alternatives = getExpressionRhsAlternatives(nextContext, lookahead);
 
         if (allowFuncCall) {
             alternatives = modifyAlternativesWithArgListStart(alternatives);
         }
         return seekInAlternativesPaths(lookahead, currentDepth, currentMatches, alternatives, isEntryPoint);
     }
 
+    private ParserRuleContext[] getExpressionRhsAlternatives(ParserRuleContext nextContext, int lookahead) {
+        ParserRuleContext[] defaultAlternatives = { ParserRuleContext.BINARY_OPERATOR,
+                ParserRuleContext.IS_KEYWORD, ParserRuleContext.DOT, ParserRuleContext.ANNOT_CHAINING_TOKEN,
+                ParserRuleContext.OPTIONAL_CHAINING_TOKEN, ParserRuleContext.CONDITIONAL_EXPRESSION,
+                ParserRuleContext.XML_NAVIGATE_EXPR, ParserRuleContext.MEMBER_ACCESS_KEY_EXPR,
+                ParserRuleContext.RIGHT_ARROW, ParserRuleContext.SYNC_SEND_TOKEN };
+
+        ParserRuleContext[] alternatives = new ParserRuleContext[defaultAlternatives.length + 1];
+
+        if (this.tokenReader.getCurrentTokenIndex() > 0) {
+            STToken previousToken = this.tokenReader.peek(lookahead - 1);
+            STNodeList trailingTrivia = (STNodeList) previousToken.trailingMinutiae();
+
+            if (!trailingTrivia.isEmpty() &&
+                    trailingTrivia.get(trailingTrivia.size() - 1).kind == SyntaxKind.END_OF_LINE_MINUTIAE) {
+                // Give higher priority to nextContext if previous token contains new line
+                // This is done to improve the recovery for missing semicolon
+                // eg:
+                // case 1 :- int a = expr1 [MISSING binaryOp] expr2;
+                // case 2 :- int a = expr1 [MISSING ;]
+                //           expr2;
+                alternatives[0] = nextContext;
+                System.arraycopy(defaultAlternatives, 0, alternatives, 1, defaultAlternatives.length);
```

Comment:
Since we frequently come here. This seems to add some overhead. Shall we hardcode two arrays instead of making one at the beginning and modifying them?